File: src/dynamic-programming/max-sum-of-rectangle-no-larger-than-k.ts
import { TreeMap } from '../utils/TreeMap'

export function maxSumSubmatrix(matrix: number[][], k: number): number {
  const rows = matrix.length
  const cols = matrix[0].length
  let ans = -Infinity

  //通过枚举矩阵上下行，然后通过累加每列的累加值可以将子矩阵和用
  //一维前缀和表示出来，这时候就能将问题简化到一维
  //这时候的问题就变为求满足presumRight - presumLeft <= k
  //不等式中(presumRight - presumLeft)的最大值,对不等式
  //做一下变换得到 presumRight - k <= presumLeft
  //结合两个不等式可以看出要求presumRight - presumLeft的最大值
  //则需要寻找最小的满足 presumRight - k <= presumLeft的presumLeft
  for (let i = 0; i < rows; ++i) {
    const arr: number[] = Array.from<number>({ length: cols }).fill(0)
    for (let j = i; j < rows; ++j) {
      for (let k = 0; k < cols; ++k) {
        arr[k] += matrix[j][k]
      }

      const map = new TreeMap<number, boolean>()
      map.put(0, true)

      let presumRight = 0

      for (const v of arr) {
        presumRight += v

        //用ceiling方法求出来的presumLeft正好满足presumLeft >= presumRight - k
        //而且还是最小的
        const presumLeft = map.ceiling(presumRight - k)

        if (presumLeft !== null) {
          ans = Math.max(ans, presumRight - presumLeft)
        }
        map.put(presumRight, true)
      }
    }
  }

  return ans
}

File: src/dynamic-programming/max-sum-of-rectangle-no-larger-than-k.c
int lower_bound(int *arr, int len, int target)
{
    int l = 0;
    int r = len - 1;

    int ans = len;
    while (l <= r)
    {
        int mid = (r - l) / 2 + l;

        if (arr[mid] == target)
        {
            ans = mid;
            r = mid - 1;
        }
        else if (arr[mid] > target)
        {
            ans = mid;
            r = mid - 1;
        }
        else
        {
            l = mid + 1;
        }
    }

    return ans;
}

void insertion_sort(int arr[], int n)
{
    int i, key, j;
    for (i = 1; i < n; i++)
    {
        key = arr[i];
        j = i - 1;
        while (j >= 0 && arr[j] > key)
        {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }
}

void print_arr(int arr[], int len)
{
    for (int i = 0; i < len; ++i)
        printf("%d, ", arr[i]);

    printf("\n");
}

int maxSumSubmatrix(int **matrix, int matrixSize, int *matrixColSize, int max)
{
    int n = matrixSize;
    int m = *matrixColSize;
    int ans = 0x80000000;
    int sum[m];
    int sorted_arr[m + 1];

    for (int i = 0; i < n; ++i)
    {
        memset(sum, 0, sizeof(int) * m);
        for (int j = i; j < n; ++j)
        {
            for (int k = 0; k < m; ++k)
            {
                sum[k] += matrix[j][k];
            }

            memset(sorted_arr, 0, sizeof(int) * (m + 1));

            sorted_arr[0] = 0;
            int len = 1;
            int presum_right = 0;

            for (int k = 0; k < m; ++k)
            {
                presum_right += sum[k];
                int presum_left_idx = lower_bound(sorted_arr, len, presum_right - max);

                if (presum_left_idx != len)
                {
                    int presum_left = sorted_arr[presum_left_idx];
                    if (presum_right - presum_left > ans)
                        ans = presum_right - presum_left;
                }

                sorted_arr[len] = presum_right;
                ++len;
                insertion_sort(sorted_arr, len);
            }
        }
    }

    return ans;
}